show-line-of-sight-between-geoelements/src/main/java/com/esri/arcgismaps/sample/showlineofsightbetweengeoelements/components/SceneViewModel.kt

```diff
@@ -0,0 +1,259 @@
+            sceneProperties.surfacePlacement = SurfacePlacement.RelativeToScene
+            graphicsOverlay.graphics.add(observerGraphic)
+            graphicsOverlay.graphics.add(taxiGraphic)
+            graphicsOverlay.renderer = renderer3D
+        }
+
+        viewModelScope.launch {
+
+            scene.load().onSuccess {
+
+                // Create a line of sight between the two graphics and add it to the analysis overlay
+                val lineOfSight = GeoElementLineOfSight(
+                    observerGeoElement = observerGraphic,
+                    targetGeoElement = taxiGraphic
+                )
+                analysisOverlay.analyses.add(lineOfSight)
+
+                // Select (highlight) the taxi when the line of sight target visibility changes to visible
+                if (lineOfSight.isVisible) {
+                    taxiGraphic.isSelected = lineOfSight.isVisible
+                }
+
+                // Create a timer to animate the tank
+                val timer = Timer()
+                timer.schedule(object : TimerTask() {
+                    override fun run() {
+                        animate()
+                    }
+                }, 0, 50)
+
+            }.onFailure { error ->
+                messageDialogVM.showMessageDialog(
+                    title = "Scene Failed to Load",
+                    description = error.cause.toString()
+                )
+
+            }
+
+
+        }
+
+    }
+
+    /**
+     * Updates [height] of the Z value
+     */
+    fun updateHeight(height: Double) {
+        val pointBuilder = PointBuilder(observerGraphic.geometry as Point).apply {
+            z = height
+        }
+        observerGraphic.geometry = pointBuilder.toGeometry()
+        _currentZValue.value = height
+    }
+
+    /**
+     * Moves the taxi toward the current waypoint a short distance.
+     */
+    private fun animate() {
+        val waypoint = wayPoints[waypointsIndex]
+        // get current location and distance from waypoint
+        var location = taxiGraphic.geometry as Point
+        val distance = GeometryEngine.distanceGeodeticOrNull(
+            location,
+            waypoint,
+            meters,
+            degrees,
+            GeodeticCurveType.Geodesic
+        )
+        // move toward waypoint a short distance
+        if (distance != null) {
+            location = GeometryEngine.tryMoveGeodetic(
+                listOf(location), 1.0, meters, distance.azimuth1, degrees,
+                GeodeticCurveType.Geodesic
+            )[0]
+        }
+        taxiGraphic.geometry = location
+
+        // rotate to the waypoint
+        if (distance != null) {
+            taxiGraphic.attributes["HEADING"] = distance.azimuth1
+        }
+
+        // reached waypoint, move to next waypoint
+        if (distance != null) {
+            if (distance.distance <= 2) {
+                waypointsIndex = (waypointsIndex + 1) % wayPoints.size
+            }
+        }
```

Comment:
- Seems like you have 3 separate null checks here for `distance`. In Kotlin, you would commonly see `?.let` being used instead of a null check

```kotlin
GeometryEngine.distanceGeodeticOrNull( .... )?.let { geodeticDistanceResult ->
    ....
}
```
- Use named arguments here for `distanceGeodeticOrNull` and `tryMoveGeodetic` to help with readability. 
- Not sure why the update for `location` is set inside of the null check, but `taxiGraphic.geometry` is being set outside of the null check, is that needed?